driver: flash: mcux_flexspi_nor: Fix for is25lpxxxd chips

JBarberU opened the pull request:
Use RDERP to detect presence of extended read register, to figure out how the chip's read register should be initialized.

The new chip revision of IS25LPXXX**D** would fail initialization as it's behavior was the same as IS25WPXXX, and thus different from IS25LPXXX.

Related to pull requests:
- #86244 
- #86369

danieldegrasse reviewed (commented):
@dleach02- this PR should probably be tested with an RT1020-EVK, or some other system with an IS25LP032 flash chip to make sure that the flash tests still pass as expected.

JBarberU committed:
driver: flash: mcux_flexspi_nor: Fix for is25lpxxxd chips

IS25LPXXXD uses the same jedec-id as IS25LPXXX, but the latter has
an extended read register, similar to IS25WPXXX. This change will
attempt to read the extended read register to determine what the
appropriate initialization value for read register should be.

Signed-off-by: John Barbero Unenge <git@lsrkttn.com>

JBarberU commented on drivers/flash/flash_mcux_flexspi_nor.c:
```diff
@@ -943,29 +943,64 @@ static int flash_flexspi_nor_sfdp_read(const struct device *dev,
+	flexspi_lut[SCRATCH_CMD][0] = FLEXSPI_LUT_SEQ(
+			kFLEXSPI_Command_SDR, kFLEXSPI_1PAD, 0x81,
+			kFLEXSPI_Command_READ_SDR, kFLEXSPI_1PAD, 0x1);
+	ret = memc_flexspi_set_device_config(&data->controller,
```
I experimented with some non-existent commands (also don't have any hardware with IS25LPXXX to test on), and the chip seems to respond with 0x00 during that testing. I implemented your suggestion to check the EB[7:4] and EB[0] to cover both bases.

danieldegrasse reviewed (dismissed):
Ok from my side, but @dleach02 this should be tested on NXP RT10xx EVKs to verify it does not impact flash driver support

de-nordic commented:
@mmahadevan108 This is NXP platform, if you do not mind I will re-assign the issue to you.

mmahadevan108 commented on drivers/flash/flash_mcux_flexspi_nor.c:
```diff
@@ -995,31 +1029,17 @@ static int flash_flexspi_nor_check_jedec(struct flash_flexspi_nor_data *data,
 		 */
-		read_params = 0;
-		ret = flash_flexspi_nor_is25_clear_read_param(data, flexspi_lut, &read_params);
+		ret = flash_flexspi_nor_is25_clear_dummy_cycles(data, flexspi_lut);
```
This case returns `-ENOTSUP`. Is this accurate given these changes made?

hakehuang commented:
test pass with flash case.
```
"testsuites"=>
  [{"name"=>"tests/drivers/flash/common/drivers.flash.common.default",
    "arch"=>"arm",
    "platform"=>"mimxrt1020_evk/mimxrt1021",
    "path"=>"tests/drivers/flash/common",
    "run_id"=>"3522388bf2cadbd2fd26ddb70fdb9835",
    "runnable"=>true,
    "retries"=>0,
    "toolchain"=>"zephyr",
    "dut"=>"0226000040214e4500121019c89200169e11000097969900",
    "status"=>"passed",
    "execution_time"=>"13.08",
    "build_time"=>"0.00",
    "testcases"=>
     [{"identifier"=>"drivers.flash.common.default.flash_driver.flash_flatten",
       "execution_time"=>"0.10",
       "status"=>"passed"},
      {"identifier"=>"drivers.flash.common.default.flash_driver.flash_copy",
       "execution_time"=>"0.07",
       "status"=>"passed"},
      {"identifier"=>"drivers.flash.common.default.flash_driver.flash_fill",
       "execution_time"=>"0.06",
       "status"=>"passed"},
      {"identifier"=>"drivers.flash.common.default.flash_driver.get_size",
       "execution_time"=>"0.00",
       "status"=>"skipped",
       "reason"=>"ztest skip"},
      {"identifier"=>
        "drivers.flash.common.default.flash_driver.flash_page_layout",
       "execution_time"=>"0.03",
       "status"=>"passed"},
      {"identifier"=>
        "drivers.flash.common.default.flash_driver.read_unaligned_address",
       "execution_time"=>"0.05",
       "status"=>"passed"},
      {"identifier"=>"drivers.flash.common.default.flash_driver.flash_erase",
       "execution_time"=>"0.07",
       "status"=>"passed"}]}]}
```

butok commented:
Hi @JBarberU 
this commit is causing a hard fault for all examples with enabled flash for **MIMXRT1160-EVK** that has **is25wp128**.

Please look at it, otherwise we have to revert it asap.

JBarberU commented:
Hey @butok! I don't have an 1160-EVK on hand to test with, but my suspicion would be that the assumptions on line 999 might not hold:

```
/*
 * Check that EB[7:4] is not all zero and that EB[0] (WIP)
 * is not 1, which should catch a chip responding to an
 * unsupported 81h command with either 0x00 or 0xFF
 */
const int has_extended_read_reg = resp_data & 0xF0 && !(resp_data & 0x01);
```

Looking at the datasheets again it seems 81h is RDERPNV for the IS25WPxxx chips vs RDERP for IS25xxxD chips (and not present on IS25LPxxx). 

Are you able to get any info on where it's failing?

butok commented:
> Hey @butok! I don't have an 1160-EVK on hand to test with, but my suspicion would be that the assumptions on line 999 might not hold:
> 
> ```
> /*
>  * Check that EB[7:4] is not all zero and that EB[0] (WIP)
>  * is not 1, which should catch a chip responding to an
>  * unsupported 81h command with either 0x00 or 0xFF
>  */
> const int has_extended_read_reg = resp_data & 0xF0 && !(resp_data & 0x01);
> ```
> 
> Looking at the datasheets again it seems 81h is RDERPNV for the IS25WPxxx chips vs RDERP for IS25xxxD chips (and not present on IS25LPxxx).
> 
> Are you able to get any info on where it's failing?

The behavior is strange. If to debug it step by step, it's fine, but as soon as to run it without stopping, it goes to z_arm_usage_fault(). I have feeling it's caused by possible wrong flash read/synchronization.

I cannot understand the motivation for the PR. But according to the PR description, it's valid only for IS25LP. 
Could you revert the code at least for IS25WP?

butok commented:
> I did notice a few discrepancies:

Most probably it will not help with this particular issue, but these settings are wrong. 
Good catch! Thank you! 

We need to review all chips, as it looks like the information was copy-pasted without correction in many places. 
May we do the PR fixes, or would you prefer to handle it yourself?
